### ISSUE TITLE
Converting circular structure to JSON     at JSON.stringify (<anonymous>)

### ISSUE THREAD

**Issue by mohamadayash** [TrackableEntities/trackable-entities-js#10]
TypeError: Converting circular structure to JSON
    at JSON.stringify (<anonymous>)
    at Request.Body.text (vendor.js?v=RCvRrqPvM2Kc5BlkEQ045FeXR6gPMRIwfn51ludN14I:72337)
    at Request.getBody (vendor.js?v=RCvRrqPvM2Kc5BlkEQ045FeXR6gPMRIwfn51ludN14I:73246)
    at Observable._subscribe (vendor.js?v=RCvRrqPvM2Kc5BlkEQ045FeXR6gPMRIwfn51ludN14I:72773)
    at Observable._trySubscribe (vendor.js?v=RCvRrqPvM2Kc5BlkEQ045FeXR6gPMRIwfn51ludN14I:245)
    at Observable.subscribe (vendor.js?v=RCvRrqPvM2Kc5BlkEQ045FeXR6gPMRIwfn51ludN14I:233)
    at FetchDataComponent.changeEntity (fetchdata.component.ts:42)
    at Object.eval [as handleEvent] (FetchDataComponent.html:24)
    at handleEvent (vendor.js?v=RCvRrqPvM2Kc5BlkEQ045FeXR6gPMRIwfn51ludN14I:23117)
    at callWithDebugContext (vendor.js?v=RCvRrqPvM2Kc5BlkEQ045FeXR6gPMRIwfn51ludN14I:24578)

**Comment by mohamadayash** [comment]
When I'm trying to post the TrackableEntity from angular app to webapi ,I've got the error above

**Comment by ydlkrishna** [comment]
I am also getting the similar error. Any update on this, please? 

CreateComponent.html:124 ERROR TypeError: Converting circular structure to JSON
    at JSON.stringify (<anonymous>)
    at HttpRequest.serializeBody (http.js:727)
    at Observable.__WEBPACK_IMPORTED_MODULE_6_rxjs_Observable__.a [as _subscribe] (http.js:1738)
    at Observable._trySubscribe (Observable.js:171)
    at Observable.subscribe (Observable.js:159)
    at subscribeToResult (subscribeToResult.js:22)
    at MergeMapSubscriber._innerSub (mergeMap.js:119)
    at MergeMapSubscriber._tryNext (mergeMap.js:116)
    at MergeMapSubscriber._next (mergeMap.js:99)
    at MergeMapSubscriber.next (Subscriber.js:86)

**Comment by tonysneed** [comment]
In the ASPNET Core Web API, there is a ConfigureServices method, where you need to configure the JSON serializer. I will post the code here in just a bit.

**Comment by ydlkrishna** [comment]
hI @tonysneed , did you get a chance to get the ASPNET Core WebAPI Config Service code to post here as mentioned above.
Meanwhile, I have tried by adding the below code but still getting the same error...

   services.AddMvc().AddJsonOptions(options =>
            {
                options.SerializerSettings.ReferenceLoopHandling = ReferenceLoopHandling.Ignore;
            });

**Comment by tonysneed** [comment]
Thanks so much. This is exactly the code I wanted to post. Cheers!

**Comment by ydlkrishna** [comment]
but it is not working out for me. still getting the same circular reference error :(

**Comment by tonysneed** [comment]
I will have a chance to try this tomorrow. Thanks for being patient.

**Comment by tonysneed** [comment]
@ydlkrishna Let me answer your question about the server-side config for JSON.Net.  I found the code snippet I used to configure the JSON serializer.

```csharp
services.AddMvc().AddJsonOptions(
        options => options.SerializerSettings.PreserveReferencesHandling = PreserveReferencesHandling.All);
```

But this won't work unless you have the .NET also on the client and you configure JSON.Net in the same way on the client.

**Comment by tonysneed** [comment]
@ydlkrishna @mohamadayash The js [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) suggest that you use a library that supports cycles, for example, [cycle.js](https://github.com/douglascrockford/JSON-js) by Douglas Crockford.  You also might want to check out [circular-json](https://github.com/WebReflection/circular-json).

The other option you have is to construct your object graph in a way that does not contain cycles.  For example, with a Product has a Category property, which in turn has a Products property, simply set the Products property to null in order to avoid the cycle.

**Comment by ydlkrishna** [comment]
@tonysneed , if I use same code without trackable entity at JS client side, it works fine. it is throwing circular reference issue only when I add trackable entity related code. In this case, how can I remove circular references ?

**Comment by tonysneed** [comment]
Yes TE on the js client side is throwing on JSON.stringify or JSON.parse. So the easiest is to remove the cycles. Can you share you object model?

**Comment by ydlkrishna** [comment]
@tonysneed , seems this error is thrown at the time of stringifying JS objects in JSON and before passing to service. So, changing the .net core Web API may not have any impact on it.

we have tried to use the Json.Stringify in your sample https://github.com/TrackableEntities/trackable-entities-js itself and able to reproduce this issue. Please find the attached app.component.ts where we tried to  [@stringify](url
[app.component.txt](https://github.com/TrackableEntities/trackable-entities-js/files/2170102/app.component.txt)

) it. 
let trackableEntity =  JSON.stringify(this.foodEntities);

**Comment by prasanna32** [comment]
@tonysneed - Please see the attached (app.component.t
[app.component.zip](https://github.com/TrackableEntities/trackable-entities-js/files/2170524/app.component.zip)
s). Inside addProduct method, When I set foodEntities.tracking to true as below, I am getting circular reference 

error.

this.foodEntities.tracking = true;
let trackableEntity =  JSON.stringify(this.foodEntities);

**Comment by tonysneed** [comment]
@prasanna32 @ydlkrishna So the error goes away when you change `TrackableSet` to just `Set`?

If you're feeling adventurous, would you be able to add a failing test to [trackable-set.spec.ts](https://github.com/TrackableEntities/trackable-entities-js/blob/master/src/trackable-set.spec.ts)?

```js
  it('should serialize object graph', (done) => {

    // Arrange
    // Act
    // Assert
    done();
  });
```

**Comment by ydlkrishna** [comment]
@tonysneed , no we meat when the value for this.foodEntities.tracking is assigned to true and using JSON.stringify(this.foodEntities); throws circular reference error. Looks like the circular reference is due to extending the js class with TracableEntity and TracableSet properties. I believe it could be coz of the adding the entity class objects to the trakableset properties that is causing circular reference. 

If we want to remove circular reference, the traking properties of framework itself should be removed. However we are giving a try with circular-json and will update the results.

meanwhile if you have any other workaround, please advise.

**Comment by tonysneed** [comment]
> when the value for this.foodEntities.tracking is assigned to true and using JSON.stringify(this.foodEntities); throws circular reference error.

This sounds like a bug — if the error does not take place with Set. So we can create a failing test, and then work to determine the exact cause.

**Comment by mohamadayash** [comment]
Thank you @tonysneed  I wish if you could fix this issue as soon as possible , regards and have a nice day

**Comment by tonysneed** [comment]
@mohamadayash Just got back home from vacation, so I'll be working on this tonight after work.  Thanks for being patient.

**Comment by tonysneed** [comment]
@mohamadayash @ydlkrishna I created a project to test `JSON.stringify` with both `Set` and `TrackableSet`:  https://github.com/tonysneed/demo-json-cycles-stringify.  Please clone the repo and run the tests.

I was not able to see any issues with `TrackableSet` with regard to serialization.  Your issues have to do with how to correctly serialize `Set` based collections.  ES6 Sets use a streaming model, so you need to convert them to arrays. See the [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) for more info.

```js
const serialized = JSON.stringify(Array.from(products.values()));
```

Secondly, you need to eliminate cycles in your object graph.  For example, you can eliminate the `products` property on `Category`.  Alternatively, you can use a replacer function, as described in the [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cyclic_object_value), but I recommend you take care to eliminate cycles in the first place, which is a better practice in my opinion.

```js
// Remove the following to eliminate cycles
category.products = Array.from(products.values()); // Causes cycles
// -- or remove Category.products altogether
```

Please let me know if this clears things up for both of you.
